lnwallet/chainfee/filtermanager.go

```diff
@@ -0,0 +1,279 @@
+	defer filterTicker.Stop()
+
+	for {
+		select {
+		case <-filterTicker.C:
+			filters, err := f.fetchFunc()
+			if err != nil {
+				log.Errorf("Encountered err while fetching "+
+					"fee filters: %v", err)
+				return
+			}
+
+			f.updateMedian(filters)
+
+		case <-f.quit:
+			return
+		}
+	}
+}
+
+// fetchMedianFilter fetches the median feefilter value.
+func (f *filterManager) fetchMedianFilter() (SatPerKWeight, error) {
+	f.medianMtx.RLock()
+	defer f.medianMtx.RUnlock()
+
+	if f.median == 0 {
+		// Return errNoFilters so the caller knows to ignore this
+		// output and continue.
+		return 0, errNoData
+	}
+
+	return f.median, nil
+}
+
+// chooseMinFee takes in the minimum relay fee and compares it against our
+// median filter, returning whichever is highest. This function does not need
+// to acquire the lock.
+func (f *filterManager) chooseMinFee(minRelayFee SatPerKWeight) SatPerKWeight {
+	filterFee, err := f.fetchMedianFilter()
+	if err == errNoData {
+		return minRelayFee
+	}
+
+	if filterFee > minRelayFee {
+		return filterFee
+	}
+
+	return minRelayFee
+}
+
+type bitcoindPeerInfoResp struct {
+	Inbound      bool    `json:"inbound"`
+	MinFeeFilter float64 `json:"minfeefilter"`
+}
+
+func fetchBitcoindFilters(client *rpcclient.Client) ([]SatPerKWeight, error) {
+	resp, err := client.RawRequest("getpeerinfo", nil)
+	if err != nil {
+		return nil, err
+	}
+
+	var peerResps []bitcoindPeerInfoResp
+	err = json.Unmarshal(resp, &peerResps)
+	if err != nil {
+		return nil, err
+	}
+
+	// We filter for outbound peers since it is harder for an attacker to
+	// be our outbound peer and therefore harder for them to manipulate us
+	// into broadcasting high-fee or low-fee transactions.
+	var outboundPeerFilters []SatPerKWeight
+	for _, peerResp := range peerResps {
+		if peerResp.Inbound {
+			continue
+		}
+
+		// The value that bitcoind returns for the "minfeefilter" field
+		// is in fractions of a bitcoin that represents the satoshis
+		// per KvB. We need to convert this fraction to whole satoshis
+		// by multiplying with COIN. Then we need to convert the
+		// sats/KvB to sats/KW.
+		//
+		// Convert the sats/KvB from fractions of a bitcoin to whole
+		// satoshis.
+		filterKVByte := SatPerKVByte(
+			peerResp.MinFeeFilter * btcutil.SatoshiPerBitcoin,
+		)
+
+		if !checkFilterBounds(filterKVByte) {
+			continue
+		}
+
+		// Convert KvB to KW and add it to outboundPeerFilters.
+		outboundPeerFilters = append(
+			outboundPeerFilters, filterKVByte.FeePerKWeight(),
+		)
+	}
+
+	// Check that we have enough data to use. We don't return an error as
+	// that would stop the querying goroutine.
+	if len(outboundPeerFilters) < minNumFilters {
+		return nil, nil
+	}
+
+	return outboundPeerFilters, nil
+}
+
+func fetchBtcdFilters(client *rpcclient.Client) ([]SatPerKWeight, error) {
+	resp, err := client.GetPeerInfo()
+	if err != nil {
+		return nil, err
+	}
+
+	var outboundPeerFilters []SatPerKWeight
+	for _, peerResp := range resp {
```

Comment:
I'd like to cite this for loop as an example of one of the many classes of for loops we can permanently eliminate if we merge #8483.

There's nothing wrong with this code, just could have condensed 186-205 into a few filter and map operations.